Sources/NIOCertificateHotReloading/NIOCertificateHotReloading.swift

```diff
@@ -0,0 +1,279 @@
+        let certificateBytes: [UInt8]?
+        switch self.certificateDescription.location._backing {
+        case .file(let path):
+            let bytes = FileManager.default.contents(atPath: path)
+            certificateBytes = bytes.map { Array($0) }
+
+        case .memory(let bytesProvider):
+            certificateBytes = bytesProvider()
+        }
+        return certificateBytes
+    }
+
+    private func loadPrivateKey() -> [UInt8]? {
+        let keyBytes: [UInt8]?
+        switch self.privateKeyDescription.location._backing {
+        case .file(let path):
+            let bytes = FileManager.default.contents(atPath: path)
+            keyBytes = bytes.map { Array($0) }
+
+        case .memory(let bytesProvider):
+            keyBytes = bytesProvider()
+        }
+        return keyBytes
+    }
+
+    private func parseCertificate(from certificateBytes: [UInt8]) -> Certificate? {
+        let certificate: Certificate?
+        switch self.certificateDescription.format._backing {
+        case .der:
+            certificate = try? Certificate(derEncoded: certificateBytes)
+
+        case .pem:
+            certificate = String(bytes: certificateBytes, encoding: .utf8)
+                .flatMap { try? Certificate(pemEncoded: $0) }
+        }
+        return certificate
+    }
+
+    private func parsePrivateKey(from keyBytes: [UInt8]) -> Certificate.PrivateKey? {
+        let key: Certificate.PrivateKey?
+        switch self.privateKeyDescription.format._backing {
+        case .der:
+            key = try? Certificate.PrivateKey(derBytes: keyBytes)
+
+        case .pem:
+            key = String(bytes: keyBytes, encoding: .utf8)
+                .flatMap { try? Certificate.PrivateKey(pemEncoded: $0) }
+        }
+        return key
+    }
+
+    private func attemptToUpdatePair(certificate: Certificate, key: Certificate.PrivateKey) {
+        let nioSSLCertificate = try? NIOSSLCertificate(
+            bytes: certificate.serializeAsPEM().derBytes,
+            format: .der
+        )
+        let nioSSLPrivateKey = try? NIOSSLPrivateKey(
+            bytes: key.serializeAsPEM().derBytes,
+            format: .der
+        )
+
+        guard let nioSSLCertificate, let nioSSLPrivateKey else {
+            return
+        }
+
+        self.state.withLockedValue {
+            $0 = CertificateKeyPair(
+                certificate: .certificate(nioSSLCertificate),
+                privateKey: .privateKey(nioSSLPrivateKey)
+            )
+        }
+    }
+}
+
+extension TLSConfiguration {
+    /// Configure a ``CertificateReloader`` to observe updates for the certificate and key pair used.
+    /// - Parameter reloader: A ``CertificateReloader`` to watch for certificate and key pair updates.
+    /// - Returns: A `TLSConfiguration` that reloads the certificate and key used in its SSL handshake.
+    mutating public func withAutomaticCertificateReloading(using reloader: any CertificateReloader) -> Self {
```

Comment:
Why are we using an existential here instead of `some`?

Reply:
Oversight :)